articles/external_qos_file.md

```diff
@@ -0,0 +1,658 @@
+
+There are currently QoS settings that are supported by DDS and not by ROS.
+An user might want to make use of them, and they won't find an easy way to do that without exposing the QoS setting on ROS.
+
+There are some cases where an user might want to use a vendor specific QoS extension, for example rti allows defining [flow controllers](https://community.rti.com/static/documentation/connext-dds/5.2.0/doc/manuals/connext_dds/html_files/RTI_ConnextDDS_CoreLibraries_UsersManual/Content/UsersManual/CreatingandConfigCustomFlowC.htm) in publishers.
+
+Though a DDS vendor specific QoS profile file can be used to set up these QoS settings, you currently can only use the default qos profile.
+There's no way to match ROS entities to the DDS QoS profile names.
+
+This problem can be solved in two ways:
+- Having conventions to match DDS QoS profiles names to ROS entities.
+  This has been proposed in https://github.com/ros2/rmw_fastrtps/pull/335, and could work well for most use cases.
+- Adding a rmw implementation specific "payload" that it's then used by the rmw implementation.
+  In DDS based implementations, this payload can be just the profile name.
+
+As an example of the second alternative:
+
+```xml
+<qos_profiles>
+    <node name="my_node" namespace="/my_ns/nested_ns">
+        <publisher topic_name="asd">
+            <qos>
+                <rmw_payload>
+                    <profile_name>my_qos_profile_name</profile_name>
+                </rmw_payload>
+            </qos>
+        </publisher>
+    </node>
+</qos_profiles>
+```
+
+```yaml
+/my_ns/nested_ns/my_node:
+    ros__qos_profiles:
+        publisher:
+            topic_name: asd
+            rmw_payload:
+                profile_name: my_qos_profile_name
+```
+
+For this to work, `rmw_qos_profile_t` will need to be extended with a `rmw_qos_profile_payload_t` member defined like:
+
+```c
+typedef struct rmw_qos_profile_payload_item_t {
+  char * key;
+  char * value;
+} rmw_qos_profile_payload_item_t;
+
+typedef struct rmw_qos_profile_payload_t {
+  rmw_qos_profile_payload_item_t * key_value_pairs;
+} rmw_qos_profile_payload_t;
+```
+
+It would be up to the implementation what to do with the passed key value pairs.
+
+Note:
+  The proposed API is limited, if we would want for example to define vendor specific QoS inline in the ROS QoS profile file.
+  If YAML format is used, we could take advantage of it's natural representation of integers,floats,etc, instead of forcing everything to be an string.
+  The complexity of this approach doesn't seem to be worth it.
+Note 2:
+  The rmw specific payload could simply be an string, instead of key value pairs.
+  In that case, the `profile_name` sub-tag would be deleted and `my_qos_profile_name` would be the content of the string.
+
+## File format
+
+As showed in the above examples, both YAML or XML can correctly represent the desired file format.
+
+XML advantages:
+- Can be easily verified.
+YAML advantages:
```

Comment:
Minor formatting issue. Need another newline before "YAML advantages.

Reply:
Ups, solved in https://github.com/ros2/design/pull/296/commits/bc3582213143ce83e8d30b515a9e799aa2a9cdc9. Thanks!